### ISSUE TITLE
TypeError: Cannot read property 'valueDeclaration' of undefined (continuation)

### ISSUE THREAD

**Issue by jasonwilliams** [microsoft/TypeScript#35801]
This is a continuation of https://github.com/microsoft/TypeScript/issues/34869

I'm still getting the error testing against `v3.8.0-dev.20191205`
Here is my stack trace (the first one is react-styleguidist)
```
Trace
    at withContext (node_modules/typescript/lib/typescript.js:36361:25)
    at Object.typeToTypeNode (node_modules/typescript/lib/typescript.js:36316:28)
    at typeToString (node_modules/typescript/lib/typescript.js:36283:40)
    at Object.typeToString (node_modules/typescript/lib/typescript.js:33355:24)
    at Parser.getDocgenType (node_modules/react-docgen-typescript/lib/parser.js:284:43)
    at node_modules/react-docgen-typescript/lib/parser.js:331:29
    at Array.forEach (<anonymous>)
    at Parser.getPropsInfo (node_modules/react-docgen-typescript/lib/parser.js:309:27)
    at Parser.getComponentInfo (node_modules/react-docgen-typescript/lib/parser.js:129:30)
    at node_modules/react-docgen-typescript/lib/parser.js:702:27
    at Array.map (<anonymous>)
    at node_modules/react-docgen-typescript/lib/parser.js:701:14
    at Array.reduce (<anonymous>)
    at parseWithProgramProvider (node_modules/react-docgen-typescript/lib/parser.js:694:10)
    at parse (node_modules/react-docgen-typescript/lib/parser.js:71:20)
    at Object.module.exports (node_modules/react-styleguidist/lib/loaders/props-loader.js:41:12)
 FAIL  Failed to compile

[internal]
INTERNAL ERROR: Cannot read property 'valueDeclaration' of undefined
stack trace:
TypeError: Cannot read property 'valueDeclaration' of undefined
    at getBaseConstructorTypeOfClass (node_modules/typescript/lib/typescript.js:40026:40)
    at serializeAsClass (node_modules/typescript/lib/typescript.js:38036:42)
    at serializeSymbolWorker (node_modules/typescript/lib/typescript.js:37770:29)
    at serializeSymbol (node_modules/typescript/lib/typescript.js:37711:38)
    at node_modules/typescript/lib/typescript.js:37686:25
    at Map.forEach (<anonymous>)
    at visitSymbolTable (node_modules/typescript/lib/typescript.js:37685:33)
    at symbolTableToDeclarationStatements (node_modules/typescript/lib/typescript.js:37556:17)
    at node_modules/typescript/lib/typescript.js:36340:106
    at withContext (node_modules/typescript/lib/typescript.js:36362:37)
error Command failed with exit code 1.
```

**Related Issues:** 
https://github.com/microsoft/TypeScript/issues/34869

**Comment by jasonwilliams** [comment]
@evenfrost  @arian @BernsteinA @elibarzilay new issue here

**Comment by RyanCavanaugh** [comment]
@jasonwilliams how do we reproduce the issue?

**Comment by jasonwilliams** [comment]
@RyanCavanaugh thats what a few of us are trying to work out, this is happening on a large private repo which isn’t useful to you so the best I can provide is a stack trace.

From the last issue we are trying to narrow this down.

**Comment by evenfrost** [comment]
I've narrowed down the issue and it appears when using `fork-ts-checker-webpack-plugin` with webpack. If I remove `ForkTsCheckerWebpackPlugin` from my webpack plugins list, build runs without errors. It happens both in dev (with `webpack-dev-server`) and production build (when using `webpack` right away).

**Comment by evenfrost** [comment]
So after 4 hours of digging I've finally tracked down the issue (for me at least). 

**TL;DR:** You should explicitly specify `tsconfig` option in `ForkTsCheckerWebpackPlugin` if you're using custom TS config file location.

Long story: I'm using `webpack`, `ts-loader`, `fork-ts-checker-webpack-plugin` and 2 separate TS config files (the one builds the client and is located in `{project_root}/client/tsconfig.json`, another one builds the server and is located in `{project_root/tsconfig.json}`). For client build I'm using `ForkTsCheckerWebpackPlugin`, which apparently has its own `tsconfig` option to specify custom `tsconfig.json` file path (which I thought would be inherited from `configFile` option of `ts-loader`). As I haven't specified `tsconfig`, `fork-ts-checker-webpack-plugin` took the config from `{project_root/tsconfig.json}`, which was server config, and `ts-loader` took it from correct location (`{project_root}/client/tsconfig.json`).

Don't know why it was working on TypeScript 3.6, but 3.7 release became really unhappy with it.

**Comment by jasonwilliams** [comment]
explicitly specifying the `tsconfig.json` location didn't help me, i still get the same error.
However, like you said, removing the ForkTsCheckerWebpackPlugin from the build removes the error.

**Comment by jasonwilliams** [comment]
I've ran git bisect against commits between `v3.6.4` (where it works) and `v3.7-rc` (where it fails). 

The offending commit seems to be this one 
https://github.com/microsoft/TypeScript/commit/29becf05012bfa7ba20d50b0d16813971e46b8a6

**Comment by evenfrost** [comment]
@jasonwilliams here's mine:

```json
{
  "extends": "./tsconfig.base.json",
  "compilerOptions": {
    "allowJs": true,
    "baseUrl": ".",
    "paths": {
      "@/*": [
        "./*"
      ]
    },
    "module": "commonjs",
    "outDir": "./build",
    "rootDir": "./",
    "target": "es2019"
  },
  "exclude": [
    "./build/**/*",
    "./client/**/*",
    "node_modules",
    "**/node_modules/*"
  ]
}
```
Content of `tsconfig.base.json`:
```json
{
  "compilerOptions": {
    "allowUnreachableCode": true,
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "noImplicitAny": true,
    "noImplicitReturns": true,
    "noImplicitThis": true,
    "noUnusedLocals": false,
    "removeComments": true,
    "sourceMap": true,
    "strictNullChecks": true,
    "suppressImplicitAnyIndexErrors": true
  }
}
```

**Comment by jasonwilliams** [comment]
@evenfrost did you try setting useTypescriptIncrementalApi: false in your typescript checker settings?

**Comment by RyanCavanaugh** [comment]
The stack trace *might* be enough to go on. We'll give it a shot.

**Comment by jasonwilliams** [comment]
@RyanCavanaugh @rbuckton  i've managed to create a test case here:
https://github.com/jasonwilliams/typescript-35801-testcase

**Comment by acomagu** [comment]
Hi there, I also created minimal set for the reproduction.

1. create *index.js* with

    ```javascript
    let A;
    A = {};
    A.prototype.b = {}
    ```
2. Run `tsc --allowJs --declaration index.js`

My tsc is v3.8.3.

I hope this helps you.

**Comment by rbuckton** [comment]
This may be related to #36273.

**Comment by rbuckton** [comment]
@weswigham I have a fix for the crash, but I'm not sure I like the JS declaration emit for @acomagu's example:

```ts
// js
let A;
A = {};
A.prototype.b = {};

// .d.ts
declare class A {
    b: {};
}
```

The binder declares `A` as both a `FunctionScopedVariable` (because of the `let`) and a `Class` because of the `prototype` assignment heuristic. The binder can't know whether its actually a class, so should we be doing some additional work in `serializeSymbolWorker` to exclude non-class-like symbols (i.e., check that it has both `SymbolFlags.Class` and some kind of call or construct signature)?

**Comment by rbuckton** [comment]
I'm thinking about something like this:

```ts
function isEffectiveClassSymbol(symbol: Symbol) {
    if (!(symbol.flags & SymbolFlags.Class)) {
        return false;
    }
    if (isInJSFile(symbol.valueDeclaration) && !isClassLike(symbol.valueDeclaration)) {
        // For a symbol that isn't syntactically a `class` in a JS file we have heuristics
        // that detect prototype assignments that indicate the symbol is *probably* a class.
        // Filter out any prototype assignments for non-class symbols, i.e.
        // 
        //     let A;
        //     A = {};
        //     A.prototype.b = {};
        const type = getTypeOfSymbol(symbol);
        return some(getSignaturesOfType(type, SignatureKind.Construct))
            || some(getSignaturesOfType(type, SignatureKind.Call));
    }
    return true;
}
```

**Comment by weswigham** [comment]
Hmm, seems reasonable.

**Comment by matthew-dean** [comment]
One note: not sure if this was stated, but I only ran into this issue with using `tsc -b`. The `tsc -p` option didn't have this issue. So (for me) it's only an issue when using the build mode.

**Comment by bartcone** [comment]
Also seeing something similar when converting a vanilla js (does use jsdoc) project over to ts `3.9.4`. 

`tsc` works, but `tsc -w` does not.

downgrading to `3.6.4` and `tsc -w` does work.

**Comment by andrewbranch** [comment]
This was fixed in #38508.
